src/common/snippets/src/runtime_configurator.cpp

```diff
@@ -272,9 +259,182 @@ void RuntimeConfigurator::update_data_offsets() const {
     }
 }
 
-void RuntimeConfigurator::update_latest_shapes() {
-    for (size_t i = 0; i < m_io_num; ++i) {
-        m_latest_shapes[i] = m_io_descs[i]->get_shape();
+std::vector<VectorDims> RuntimeConfigurator::extract_shapes() {
+    std::vector<VectorDims> shapes(m_io_num);
+    for (size_t i = 0; i < m_io_num; ++i)
+        shapes[i] = m_io_descs[i]->get_shape();
+    return shapes;
+}
+
+std::vector<std::vector<size_t>> RuntimeConfigurator::extract_layouts() {
+    std::vector<std::vector<size_t>> layouts(m_io_num);
+    for (size_t i = 0; i < m_io_num; ++i)
+        layouts[i] = m_io_descs[i]->get_layout();
+    return layouts;
+}
+
+RuntimeConfigurator::UnifiedLoopInfoRtParams RuntimeConfigurator::compute_runtime_params(const lowered::UnifiedLoopInfoPtr& loop_info) {
+    RuntimeConfigurator::UnifiedLoopInfoRtParams rt_params;
+    rt_params.work_amount = loop_info->get_work_amount();
+    const auto count = loop_info->get_input_count() + loop_info->get_output_count();
+    rt_params.ptr_increments.resize(count);
+    rt_params.finalization_offsets.resize(count);
+
+    size_t idx = 0;
+    loop_info->iterate_through_descs(
+        [&rt_params, &idx](const lowered::UnifiedLoopInfo::LoopPortDesc& desc) {
+            rt_params.ptr_increments[idx] = desc.ptr_increment;
+            rt_params.finalization_offsets[idx] = desc.finalization_offset;
+            ++idx;
+        });
+    return rt_params;
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::init(const lowered::LinearIRPtr& linear_ir,
+                                                    const std::vector<PortDescriptorPtr>& io_descs,
+                                                    size_t in_num) {
+    if (linear_ir->get_config().m_enable_domain_optimization || !linear_ir->is_dynamic())
+        return;
+    concurrency = linear_ir->get_config().m_min_parallel_work_amount;
+
+    const auto brgemms = find_applicable_brgemms(linear_ir);
+    // Parallel WA optimization is Brgemm related
+    if (brgemms.empty())
+        return;
+
+    init_non_m_related_params(linear_ir, brgemms);
+    OPENVINO_ASSERT(!not_m_related_params.empty(), "not_m_related_params mustn't be empty after initialization");
+    init_loops_to_split(linear_ir);
+
+    m_dim_idces.resize(io_descs.size());
+    split_layouts.resize(io_descs.size());
+    for (size_t i = 0; i < io_descs.size(); ++i) {
+        const auto& layout = io_descs[i]->get_layout();
+        const auto dim_idx = i < in_num ? utils::get_input_dim_idx(layout, 1)
+                                        : utils::get_output_dim_idx(layout, 1);
```

Comment:
To make this optimization applicable for other Subgraph patterns in the future, you can replace `dim_idx` with `m_config->tile_rank-1` and `2` with `m_config->tile_rank` whenever possible

Reply:
I am not sure that we can replace it with the proposed values: the optimization aims on exactly M dimension of brgemm ops, I think we have to concentrate on that.

Reply:
Again, it will be great if we have some basement for the further generic algorithm (not for MHA).
But I don't insist, just a suggestion 😃